RecoParticleFlow/PFClusterProducer/plugins/alpaka/PFClusterSoAProducerKernel.dev.cc

```diff
@@ -0,0 +1,1652 @@
+          // PF cluster is isolated seed. No iterations needed
+          if (alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] == 0) {
+            // Fill PFCluster-level information
+            int rhIdx = pfClusteringVars[pfClusteringVars[topoId].topoSeedOffsets()]
+                            .topoSeedList();  // i is the seed rechit index
+            int seedIdx = pfClusteringVars[rhIdx].rhIdxToSeedIdx();
+            clusterView[seedIdx].energy() = pfRecHits[rhIdx].energy();
+            clusterView[seedIdx].x() = pfRecHits[rhIdx].x();
+            clusterView[seedIdx].y() = pfRecHits[rhIdx].y();
+            clusterView[seedIdx].z() = pfRecHits[rhIdx].z();
+          }
+        } else if (nSeeds <= 400 && nRHTopo - nSeeds <= 1500) {
+          dev_hcalFastCluster_multiSeedIterative(
+              acc, pfClusParams, topoId, nSeeds, nRHTopo, pfRecHits, pfClusteringVars, clusterView, fracView);
+        } else {
+          if (alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u] == 0)
+            printf("Topo cluster %d has %d seeds and %d rechits. Will be processed in next kernel.\n",
+                   topoId,
+                   nSeeds,
+                   nRHTopo);
+        }
+      } else
+        return;
+    }
+  };
+
+  PFClusterProducerKernel::PFClusterProducerKernel(Queue& queue, const reco::PFRecHitHostCollection& pfRecHits)
+      : nSeeds(cms::alpakatools::make_device_buffer<uint32_t>(queue)),
+        globalClusterPos(cms::alpakatools::make_device_buffer<Position4[]>(queue, 4 * 2304)),
+        globalPrevClusterPos(cms::alpakatools::make_device_buffer<Position4[]>(queue, 4 * 2304)),
+        globalClusterEnergy(cms::alpakatools::make_device_buffer<float[]>(queue, 4 * 2304)),
+        globalRhFracSum(cms::alpakatools::make_device_buffer<float[]>(queue, 4 * 2304)),
+        globalSeeds(cms::alpakatools::make_device_buffer<int[]>(queue, 4 * 2304)),
+        globalRechits(cms::alpakatools::make_device_buffer<int[]>(queue, 4 * 2304)) {
+    alpaka::memset(queue, nSeeds, 0x00);  // Reset nSeeds
+  }
+
+  void PFClusterProducerKernel::execute(const Device& device,
+                                        Queue& queue,
+                                        const reco::PFClusterParamsDeviceCollection& params,
+                                        reco::PFClusteringVarsDeviceCollection& pfClusteringVars,
+                                        reco::PFClusteringEdgeVarsDeviceCollection& pfClusteringEdgeVars,
+                                        const reco::PFRecHitHostCollection& pfRecHits,
+                                        reco::PFClusterDeviceCollection& pfClusters,
+                                        reco::PFRecHitFractionDeviceCollection& pfrhFractions) {
+    const int nRH = pfRecHits->size();
+    const int threadsPerBlock = std::is_same_v<Device, alpaka::DevCpu> ? 32 : 256;
+    const int blocks = std::is_same_v<Device, alpaka::DevCpu> ? nRH : (nRH + threadsPerBlock - 1) / threadsPerBlock;
```

Comment:
As it's used very often we have a wrapper for this construct.
You can replace `(nRH + threadsPerBlock - 1) / threadsPerBlock` with `divide_up_by(nRH, threadsPerBlock)`.